### PR TITLE
End-to-end tests using Jasmine + Webdriver

### PR DESCRIPTION
**Pull request by tlrobinson**
Implements #689

### CODE REVIEW

**Comment by tlrobinson** [comment]
How do we want to integrate this into CI? Obviously since it's end-to-end it requires both the frontend and backend to be built, which can take a fairly long time. @salsakran has mentioned wanting to use jars built on CI in our CD process (which currently uses Dockerhub to build separately), so if we do want to do that we'll have to build the jar anyway. This would likely end up being the bottleneck Circle node (though we might also consider reducing the parallelism such that we can do two full tests at once, maybe 4 nodes each)

We also need to figure out how to set up Selenium on Circle, or point it to Sauce labs or similar.

**Comment by lewisl9029** [comment]
I think one option might be to make use of CircleCI's own Docker integration: https://circleci.com/docs/docker/

Inside the CircleCI node, we can build a Docker container to run tests and build the jar (inside the same container), and then push it to Docker Hub to deploy. Or if we want to deploy a container with a smaller footprint, we can expose the jar from the test container and build a new minimal container with only the jar and its dependencies and push that for deployment separately. Though we might still want to push the test container to Docker Hub as well and pull it in later builds for caching purposes, depending on how well CircleCI's [officially recommended caching solution](https://circleci.com/docs/docker/#caching-docker-layers) works.

Setting up Selenium with Docker should be pretty straightforward as well, just a matter of pulling the official Selenium Docker image we want to run against and writing a docker-compose to hook up the containers.

This sounds like a good plan in my head right now but might be too naive depending on how our CI process is currently set up. Any thoughts?

**Comment by tlrobinson** [comment]
I cleaned this up and wrote a couple auth tests. However this is blocked on #3115 since it includes a db committed to the repo, which has my jar's path set as the sample db path. @camsaul I'll need some help coming up with a solution or workaround (copy the jar to a known location like `/tmp`?). Alternatively I can just run the e2e tests for now...

**Comment by tlrobinson** [comment]
To attach some real numbers to this, a full frontend+backend+uberjar build plus ~5 simple e2e tests is taking about 15 minutes, or twice as long as our current longest CI node.

I think we could speed that up quite a bit by:

1. caching `node_modules`. For some reason we're not doing `npm install` as part of Circle's dependencies phase, probably because it slows down the nodes that don't need it, but as a consequence we can't cache the `node_modules` directory. I believe we also ran into weird non-deterministic `npm install` issues (yay!) but those seem to be rare these days, and can be addressed with the "Rebuild without cache" button.
2. disabling minification. On my machine this cuts the webpack build time in half

**Comment by tlrobinson** [comment]
Here's a list of manual tests we do as part of the release process which we should try to automate using this:

1. Start a fresh install of Metabase and run through the Setup process
    a. Check that you can add a database during setup.
    b. Make sure that you are automatically showing our modal onboarding screens when setup stages are complete.
    c. Go through the QB tutorial real quick and make sure things look normal.
2. Add some users to the installation
    a. Set the role for the user
    b. Check that you can reset a user’s password
    c. Update a user’s details
3. Edit the data model for your test database
    a. Hide a table
    b. Set field visibility type
    c. Set field special type
    d. Create a segment
    e. Create a metric
4. Use the QB to create and save some questions
    a. Pivot tables
    b. Charts
    c. SQL/Native queries
    d. Use metrics and segments
    e. Use custom fields
5. Build a dashboard out of your new questions
    a. Resizing and repositioning of cards
    b. Adding series to cards
    c. Full screen mode
    d. Auto refresh
6. Create a pulse and send it to a couple people

**Comment by tlrobinson** [comment]
@salsakran @camsaul I think this is ready to go. It's got the e2e tests passing (some of them are disabled until I have a chance to make them more reliable) and disables some of the tests by default, so normal build times should be down to about 6 minutes.

The jar building (`[ci jar]`), e2e tests (`[ci e2e]`), screenshot comparison (`[ci screenshots]`), and individual driver tests (`[ci drivers]`) are now disabled by default. You can enable them individually or all of them with `[ci all]`. They are all always run on the `master` branch.

I also moved one of the backend linters from node 4 to 5 so those two are more balanced, since they're the bottleneck when not running the drivers or e2e tests.

@camsaul can you take a look at `/bin/ci` and make sure I didn't do anything too crazy?

File: frontend/test/e2e/admin/datamodel.spec.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            await waitForElementAndClick(driver, ".ColumnarSelector-rows li:nth-child(2) .ColumnarSelector-row");
+
+            await waitForElementAndClick(driver, "#ColumnsList li:nth-child(2) .TableEditor-field-visibility");
+            await waitForElementAndClick(driver, ".ColumnarSelector-rows li:nth-child(3) .ColumnarSelector-row");
+
+            // modify special type for address field
+            await waitForElementAndClick(driver, "#ColumnsList li:first-child .TableEditor-field-special-type");
+            await waitForElementAndClick(driver, ".ColumnarSelector-rows li:nth-child(2) .ColumnarSelector-row");
+
+            //TODO: verify tables and fields are hidden in query builder
```
Comment (salsakran) [inline frontend/test/e2e/admin/datamodel.spec.js]: Was this a TODO from the Lewis PR or are we still waiting on this? Is it easy to verify?

Either way, this doesn't block merging for me.

**Review by salsakran — APPROVED** [review]
Overall looks good, but I didn't go over the tests too closely/

File: frontend/test/e2e/admin/datamodel.spec.js (hunk 1 above)
Comment (tlrobinson) [inline frontend/test/e2e/admin/datamodel.spec.js]: Yeah I haven't finished Lewis' tests, I just disabled the ones that were still failing. I'd rather just get this in so we can iterate on it, but I can split the disabled/incomplete ones into a separate PR if you want.

File: frontend/src/metabase/lib/card.js [hunk 2]
```diff
@@ -66,6 +66,7 @@ export function isCardDirty(card, originalCard) {
 }
 
 export function serializeCardForUrl(card) {
+    // console.log(JSON.stringify(card, null, '  '));
```
Comment (kdoh) [inline frontend/src/metabase/lib/card.js]: Looks like this is something we can get rid of.

File: frontend/src/metabase/meta/types/Database.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/* @flow */
+
+import type { Table } from "./Table";
+
+export type DatabaseId = number;
+
+// TODO: incomplete
+export type Database = {
+    id: DatabaseId,
+
```
Comment (kdoh) [inline frontend/src/metabase/meta/types/Database.js]: Hey, nitpicky Kyle here. but any reason for the space?

**Review by kdoh — COMMENTED** [review]
Overall looks pretty good. I'm a bit curious about why we're adding so many IDs in this PR though. I get why we'd use them for identifiers in testing in the abstract, but I wasn't able to get a clear sense of how often they're used in the actual test code, and if they have any performance benefits in this context.

**Comment by tlrobinson** [comment]
@kdoh Yeah, I'm not a huge fan of adding ids just for tests. I've been trying to think of a way to select nodes by their React component name, and think I just came up with a decent (though kind of hacky) solution: https://jsfiddle.net/tlrobinson/wsb43qub/

Basically in tests we can iterate over every DOM node and add each React component's name as an attribute on the component's root DOM node (e.x. `<div react="Dashboard">...`), then use attribute CSS selectors to do something like `document.querySelector("[react=Dashboard] [react=NightModeIcon]").click()`.

We still may need to occasionally add an extra attribute (I think I'd prefer `class` or `data-` attribute over `id`) but I think the need will be greatly reduced.

I'm ok with delaying merging this until we see if this works and strip out those ids.

**Comment by kdoh** [comment]
@tlrobinson I'm not so hung up on the ID thing and I think more tests > perfect tests so feel free to merge this whenever. Your solution looks pretty cool though 👍

File: frontend/test/e2e/query_builder/query_builder.spec.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            await waitForElementRemoved(driver, ".EditHeader");
+        });
+    });
+
+    describe("charts", () => {
+        xit("should allow users to create line charts", async () => {
+            await driver.get(`${server.host}/q`);
+
+            // select orders table
+            await waitForElementAndClick(driver, "#TablePicker .List-item:first-child>a");
```
Comment (salsakran) [inline frontend/test/e2e/query_builder/query_builder.spec.js]: hmm. this feels fragile (specifically that the orders table is the first child of the table picker.)

**Comment by salsakran** [comment]
See a few things that make me itchy, but getting an e2e test suite merged and letting it improve with time makes me a very strong 👍  despite any nits I have to pick.

**Pull request merged**